### ISSUE TITLE
JSON stringify issue

### ISSUE THREAD

**Issue by yet1moreed** [pocoproject/poco#1042]
Greetings,

It appears that Poco JSON Stringifier does not handle all the control codes in the range 0x00 to 0x1F.  According to RFC 4627 section 2.5, all codes in this range must be escaped. While some codes in the range are escaped, any that require /u (Unicode hex encoding) to do so, not are not escaped but instead are injected directly into the output stream.

RFC 4627 Section 2.5:

 The representation of strings is similar to conventions used in the C
   family of programming languages.  A string begins and ends with
   quotation marks.  All Unicode characters may be placed within the
   quotation marks except for the characters that must be escaped:
   quotation mark, reverse solidus, and the control characters (U+0000
   through U+001F).

By not escaping such control codes, the resulting JSON output may not be valid.  

Below is such an example where the control code 0x17 is properly parsed using the Poco JSON Parser, but when using stringify, the result is invalid JSON.  The string in "result" does not quote the control character (0x17) as required by the specification.

```
    std::stringstream s;
    std::string json( "[ \"\\u0017\" ]" );
    Poco::JSON::Parser parser;
    Poco::Dynamic::Var v = parser.parse( json );
    Poco::JSON::Stringifier::condense( v, s );
    std::string result = s.str();
```

Poco 1.6.1 was used to test the example code.

-Ed

**Comment by aleks-f** [comment]
Should be added to [this pull](https://github.com/pocoproject/poco/pull/1027).

**Comment by aleks-f** [comment]
I could be missing something here, but this problem was either fixed or never existed. I can not reproduce it:

```
std::stringstream s;
json = "[ \"\\u0017\" ]";
Var v = Parser().parse(json);
Stringifier::condense(v, s);
assert(s.str() == "[\"\\u0017\"]");
```

works fine for me.

**Comment by yet1moreed** [comment]
I checked this again and downloaded the latest stable source code (1.7.2).  The hasn't been changed and only has special handling for some control characters.  See below.  If there is anything I can do to further demonstrate please let me know.

-Ed 



void Stringifier::formatString(const std::string& value, std::ostream& out)
{
	out << '"';
	for (std::string::const_iterator it = value.begin(),
		 end = value.end(); it != end; ++it)
	{
		switch (*it)
		{
			case '\\': out << "\\\\"; break;
			case '"': out << "\\\""; break;
			case '/': out << "\\/"; break;
			case '\b': out << "\\b"; break;
			case '\f': out << "\\f"; break;
			case '\n': out << "\\n"; break;
			case '\r': out << "\\r"; break;
			case '\t': out << "\\t"; break;
			default: out << *it; break;
		}
	}
	out << '"';
}

**Comment by aleks-f** [comment]
Actually, it is [fixed](https://github.com/pocoproject/poco/commit/beb81b4f1684122f51f7083123093ff579e0996b#diff-128f2c58cc4a19fc5c01453a18706d06L70) in develop and will be released in 1.8. It just was never backported to 1.6 and 1.7 branches.

**Comment by yet1moreed** [comment]
Super.  Many thanks.
-Ed

**Comment by berkus** [comment]
It does NOT seem fixed in [beb81b4f1684122f51f7083123093ff579e0996b]

Try adding `assert (toJSON('\v') == "\\u000b");` and you'll see it doesn't work.

**Comment by aleks-f** [comment]
toJSON is part of Foundation, not JSON library so it's a separate issue
